stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/decorator/internal/AvatarDecorator.kt

```diff
@@ -2,38 +2,77 @@ package io.getstream.chat.android.ui.message.list.adapter.viewholder.decorator.i
 import com.getstream.sdk.chat.adapter.MessageListItem
-import com.getstream.sdk.chat.utils.extensions.isBottomPosition
 import io.getstream.chat.android.ui.avatar.AvatarView
+import io.getstream.chat.android.ui.message.list.DefaultShowAvatarPredicate
+import io.getstream.chat.android.ui.message.list.MessageListView
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.GiphyViewHolder
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.MessageDeletedViewHolder
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.MessagePlainTextViewHolder
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.internal.TextAndAttachmentsViewHolder
 
-internal class AvatarDecorator : BaseDecorator() {
+internal class AvatarDecorator(
+    private val showAvatarPredicate: MessageListView.ShowAvatarPredicate = DefaultShowAvatarPredicate(),
+) : BaseDecorator() {
     override fun decorateTextAndAttachmentsMessage(
         viewHolder: TextAndAttachmentsViewHolder,
         data: MessageListItem.MessageItem,
     ) {
-        setupAvatar(viewHolder.binding.avatarView, data)
+        setupAvatar(getAvatarView(viewHolder, data.isMine), data)
     }
 
     override fun decoratePlainTextMessage(
         viewHolder: MessagePlainTextViewHolder,
         data: MessageListItem.MessageItem,
     ) {
-        setupAvatar(viewHolder.binding.avatarView, data)
+        setupAvatar(getAvatarView(viewHolder, data.isMine), data)
     }
 
     override fun decorateGiphyMessage(
         viewHolder: GiphyViewHolder,
         data: MessageListItem.MessageItem,
     ) = Unit
 
+    override fun decorateDeletedMessage(viewHolder: MessageDeletedViewHolder, data: MessageListItem.MessageItem) {
+        setupAvatar(getAvatarView(viewHolder, data.isMine), data)
+    }
+
     private fun setupAvatar(avatarView: AvatarView, data: MessageListItem.MessageItem) {
-        if (data.isTheirs && data.isTheirs && data.isBottomPosition()) {
-            avatarView.isVisible = true
+        val shouldShow = showAvatarPredicate.shouldShow(data)
+
+        avatarView.isVisible = shouldShow
+
+        if (shouldShow) {
             avatarView.setUserData(data.message.user)
+        }
+    }
+
+    private fun getAvatarView(holder: TextAndAttachmentsViewHolder, isMine: Boolean): AvatarView {
+        return if (isMine) {
+            holder.binding.avatarView.isVisible = false
+            holder.binding.avatarMineView
+        } else {
+            holder.binding.avatarMineView.isVisible = false
+            holder.binding.avatarView
+        }
+    }
+
+    private fun getAvatarView(holder: MessagePlainTextViewHolder, isMine: Boolean): AvatarView {
+        return if (isMine) {
+            holder.binding.avatarView.isVisible = false
+            holder.binding.avatarMineView
+        } else {
+            holder.binding.avatarMineView.isVisible = false
+            holder.binding.avatarView
+        }
+    }
```

Comment:
This might be worth de-duplicating, as we're doing the same thing three times here

Reply:
The binding views are not the same, so I don't think it is possible to make a method for all of then

Reply:
I'm thinking something like `getAvatarView(myView: AvatarView, theirView: AvatarView, isMine: Boolean)`, we have similar things in some of our decorators